sql-plugin/src/main/scala/org/apache/spark/sql/rapids/predicates.scala

```diff
@@ -70,63 +65,52 @@ case class GpuAnd(left: Expression, right: Expression) extends CudfBinaryOperato
 
   override def binaryOp: BinaryOp = BinaryOp.LOGICAL_AND
 
-  import GpuLogicHelper._
-
-  // The CUDF implementation of `and` will return a null if either input is null
-  // Spark does not.
-  // |LHS/RHS| TRUE  | FALSE | NULL  |
-  // |TRUE   | TRUE  | FALSE | NULL  |
-  // |FALSE  | FALSE | FALSE | FALSE |
-  // |NULL   | NULL  | FALSE | NULL  |
-  // So we have to make some adjustments.
-  //  IF (A <=> FALSE, FALSE,
-  //    IF (B <=> FALSE, FALSE,
-  //      A cudf_and B))
-
   override def doColumnar(lhs: GpuColumnVector, rhs: GpuColumnVector): ColumnVector = {
     val l = lhs.getBase
     val r = rhs.getBase
-    withResource(Scalar.fromBool(false)) { falseVal =>
-      val firstPass = withResource(l.and(r)) { lAndR =>
-        withResource(l.equalToNullAware(falseVal)) { lIsFalse =>
-          lIsFalse.ifElse(falseVal, lAndR)
-        }
-      }
-      withResource(firstPass) { firstPass =>
-        withResource(r.equalToNullAware(falseVal)) { rIsFalse =>
-          rIsFalse.ifElse(falseVal, firstPass)
+    withResource(new Table(l, r)) {
+      table => {
+        val astOp = new ast.BinaryOperation(ast.BinaryOperator.NULL_LOGICAL_AND,
+          new ast.ColumnReference(0),
+          new ast.ColumnReference(1))
+        withResource(astOp.compile()) { compiledAstExpr =>
+          compiledAstExpr.computeColumn(table)
         }
       }
     }
   }
 
   override def doColumnar(lhs: GpuScalar, rhs: GpuColumnVector): ColumnVector = {
-    val l = lhs.getBase
     val r = rhs.getBase
-    withResource(Scalar.fromBool(false)) { falseVal =>
-      if (eqNullAware(lhs, false)) {
-        ColumnVector.fromScalar(falseVal, r.getRowCount.toInt)
-      } else {
-        withResource(l.and(r)) { lAndR =>
-          withResource(r.equalToNullAware(falseVal)) { rIsFalse =>
-            rIsFalse.ifElse(falseVal, lAndR)
-          }
+    val lhsLit = if (lhs.isValid) {
+      ast.Literal.ofBoolean(lhs.getValue.asInstanceOf[Boolean])
```

Comment:
It's mildly unfortunate we don't have a way to directly pass the cudf `Scalar` directly to the AST, as this is unnecessary GPU synchronization (both for the `isValid` and the `getValue`).  Given that ultimately when the AST is compiled it generates a cudf `Scalar` for the literal, (with corresponding GPU synchronization there), it's a lot of unnecessary (if likely very small) synchronizations.  I'll file a cudf issue to support passing cudf `Scalar` instances to an AST directly.

Reply:
Filed rapidsai/cudf#10002 for this.